File: 03-backtracking-generation/generation-2.py
def init(a, k):
    for i in range(k):
        a.append(i + 1)

def out(a, k):
    print(" ".join(map(str, a[:k])))

def gen(a, k, n):
    i = k - 1
    while i >= 0 and a[i] == n - k + i + 1:
        i -= 1

    if i >= 0:
        a[i] += 1
        for j in range(i + 1, k):
            a[j] = a[i] + j - i

def is_last(a, k, n):
    for i in range(k):
        if a[i] != n - k + i + 1:
            return False
    return True

def method(a, n, k):
    init(a, k)
    out(a, k)
    while not is_last(a, k, n):
        gen(a, k, n)
        out(a, k)

def main():
    n = int(input("Input n: "))
    k = int(input("Input k: "))
    a = []
    method(a, n, k)

if __name__ == "__main__":
    main()

File: 03-backtracking-generation/generation-2.cpp
#include <stdio.h>
#include <iostream>

using namespace std;

void init(int a[], int k) {
    for (int i = 0; i < k; i++) {
        a[i] = i + 1;
    }
}

void out(int a[], int k) {
    for (int i = 0; i < k; i++) {
        cout << a[i] << " ";
    }
    cout << endl;
}

void gen(int a[], int k, int n) {
    int i = k - 1;
    while (i >= 0 && a[i] == n - k + i + 1) {
        i--;
    }

    if (i >= 0) {
        a[i]++;
        for (int j = i + 1; j < k; j++) {
            a[j] = a[i] + j - i;
        }
    }
}

int islast(int a[], int k, int n) {
    for (int i = 0; i < k; i++) {
        if (a[i] != n - k + i + 1) return 0;
    }
    return 1;
}

void method(int a[], int n, int k) {
    init(a, k);
    out(a, k);
    int stop = islast(a, k, n);
    while (stop == 0) {
        gen(a, k, n);
        out(a, k);
        stop = islast(a, k, n);
    }
}

int main() {
    int k, n;
    cout << "Input n: ";
    cin >> n;
    cout << "Input k: ";
    cin >> k;
    int a[k];
    method(a, n, k);
    return 0;
}